indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -109,34 +124,150 @@ public static void setupClasspath(
+                  } else {
+                    addJarToClassPath(jarFile, distributedClassPath, intermediateClassPath, fs, conf);
                   }
+                  return true;
                 }
-            );
-          }
+              },
+              shouldRetryPredicate(),
+              DEFAULT_RETRY_COUNT
+          );
+        } catch (Exception e){
+          throw Throwables.propagate(e);
+        }
+      }
+    }
+  }
 
-          existing.add(hdfsPath);
+  public static final Predicate<Throwable> shouldRetryPredicate()
+  {
+    return new Predicate<Throwable>()
+    {
+      @Override
+      public boolean apply(Throwable input)
+      {
+        if (input == null) {
+          return false;
         }
+        if (input instanceof IOException) {
+          return true;
+        }
+        return apply(input.getCause());
+      }
+    };
+  }
 
-        DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
+  private static void addJarToClassPath(
+      File jarFile,
+      Path distributedClassPath,
+      Path intermediateClassPath,
+      FileSystem fs,
+      Configuration conf
+  )
+      throws IOException
+  {
+    // Non-snapshot jar files are uploaded to the shared classpath.
+    final Path hdfsPath = new Path(distributedClassPath, jarFile.getName());
+    if (!fs.exists(hdfsPath)) {
+      // Muliple jobs can try to upload the jar here,
+      // to avoid them from overwriting files, first upload to intermediateClassPath and then rename to the distributedClasspath.
+      final Path intermediateHdfsPath = new Path(intermediateClassPath, jarFile.getName());
+      uploadJar(jarFile, intermediateHdfsPath, fs);
+      IOException exception = null;
+      try {
+        log.info("Renaming jar to path[%s]", hdfsPath);
+        fs.rename(intermediateHdfsPath, hdfsPath);
+        if (!fs.exists(hdfsPath)) {
+          throw new ISE("File does not exist even after moving from[%s] to [%s]", intermediateHdfsPath, hdfsPath);
+        }
+      }
+      catch (IOException e) {
+        // rename failed, possibly due to race condition. check if some other job has uploaded the jar file.
+        try {
+          if (!fs.exists(hdfsPath)) {
+            log.error(e, "IOException while Renaming jar file");
+            exception = e;
+          }
+        }
+        catch (IOException e1) {
+          e.addSuppressed(e1);
+          exception = e;
+        }
+      }
+      finally {
+        try {
+          if (fs.exists(intermediateHdfsPath)) {
+            fs.delete(intermediateHdfsPath, false);
+          }
+        }
+        catch (IOException e) {
+          if (exception != null) {
+            exception.addSuppressed(e);
+          } else {
+            exception = e;
+          }
+        }
+        if(exception != null){
+          throw exception;
+        }
       }
     }
+    DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
```

Comment:
DistributedCache is deprecated, can you use `job.addFileToClassPath` to add the files?

Yes, right now that seems to do exactly this under the hood, but if hadoop ever gets their crap together they can change this without warning.